<h1>Loading Necessary Libraries</h1>

In [373]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

<h2>Data PreProcessing Starts</h2>

In [374]:
df=pd.read_csv('test.csv')
df_survived=pd.read_csv('gender_submission.csv')

In [375]:
df.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [376]:
df_survived.head()

PassengerId  Survived
0          892         0
1          893         1
2          894         0
3          895         0
4          896         1

# Concating df and df_survived because df doesnot have survived column and both are aligned perfectly


In [377]:
df=pd.concat([df,df_survived['Survived']],axis=1)

In [378]:
df.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  Survived  
0  34.5      0      0   330911   7.8292   NaN        Q         0  
1  47.0      1      0   363272   7.0000   NaN        S         1  
2  62.0      0      0   240276   9.6875   NaN        Q         0  
3  27.0      0      0   315154   8.6625   NaN        S         0  
4  22.0      1      1  3101298  12.2875   NaN        S         1

In [379]:
df.sample(5)

PassengerId  Pclass                                        Name     Sex  \
402         1294       1              Gibson, Miss. Dorothy Winifred  female   
165         1057       3  Kink-Heilmann, Mrs. Anton (Luise Heilmann)  female   
273         1165       3                          Lennon, Miss. Mary  female   
416         1308       3                         Ware, Mr. Frederick    male   
274         1166       3                       Saade, Mr. Jean Nassr    male   

      Age  SibSp  Parch  Ticket    Fare Cabin Embarked  Survived  
402  22.0      0      1  112378  59.400   NaN        C         1  
165  26.0      1      1  315153  22.025   NaN        S         1  
273   NaN      1      0  370371  15.500   NaN        Q         1  
416   NaN      0      0  359309   8.050   NaN        S         0  
274   NaN      0      0    2676   7.225   NaN        C         0

In [380]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
 11  Survived     418 non-null    int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 39.3+ KB


In [381]:
df.shape

(418, 12)

In [382]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
 11  Survived     418 non-null    int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 39.3+ KB


# We have 418 total datas and in some we can observe value for nulls. 
Checking for null values for each columns

In [383]:
df.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
Survived         0
dtype: int64

# Determining how to fill datas for age


In [384]:
print('Oldest Passenger was of:',df['Age'].max(),'Years')
print('Youngest Passenger was of:',df['Age'].min(),'Years')
print('Average Age on the ship:',df['Age'].mean(),'Years')

#So maybe mean would be appropriate

Oldest Passenger was of: 76.0 Years
Youngest Passenger was of: 0.17 Years
Average Age on the ship: 30.272590361445783 Years



# Determining how to fill datas for Cabins


In [385]:
print(len(df['Cabin'].unique()))

#Many unique Values. 

print(df['Cabin'].sample(10))

#Values seem not appropiate to take mean.


77
52                 NaN
176                NaN
11                 NaN
18                 NaN
142    B57 B59 B63 B66
80                 NaN
57               F G63
252                C89
319                NaN
145                NaN
Name: Cabin, dtype: object


In [386]:

# Checking Cabin's Relevance to output


In [387]:

#df['Cabin'].corr(df['Survived']) Couldnot be computed

#Also if we analyze Cabin will have no importance to survive rate

#Best to drop them.

# Determining how to fill datas for Fare


In [388]:
print(len(df['Fare'].unique()))

#Has 170 unique Values. 
#So a Mean would be appropiate


170


## Lets Fill NAN Feilds

In [389]:
df['Fare'].fillna(df['Fare'].mean(),inplace=True) #Replacing with mean values
df['Age'].fillna(df['Age'].mean(),inplace=True) #Replacing with mean values
df.drop('Cabin',axis=1,inplace=True)

In [390]:
df.columns

Index(['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Embarked', 'Survived'],
      dtype='object')

Age has 177 null values
Cabin has 687 null values
Emabrked has 2 null values

We can drop them. But 177 and 687 are too large numbers to drop.
Further preprocessing might be required.

# A Function to view more details with Survival

In [391]:

def Detailed_Details(df,target,outcome):
    print("Comparing",target," With",outcome)
    print("Total",len(df))
    print("\n")
    print("Column Name\t\t\tTotal\tPercentage\tTotal_Outcome(1)\tPercentage_Outcome(1)\tTotal_Outcome(0)\tPercentage_Outcome(0) ")
    #for checking type
    type={'int','int16','int32','int64','float','float16','float32','float64'}
   
    uniq_cases=len(df)/5
    total_data=len(df)
   
    df_current=df[target]
    check_type=df_current.dtypes
    
    #for formatting
    design=0
    
    if(str(check_type) in type):
    
        if (df_current.unique().sum()<uniq_cases):
            
            uniq_sorted=sorted(df_current.unique())
            for uniq in uniq_sorted:
                
                total_outcome_1=0
                total_outcome_0=0
                for j in range(0,len(df)):
                    if((df[outcome][j]==1) and (df_current[j]==uniq)):
                        total_outcome_1+=1
                    elif((df[outcome][j]==0) and (df_current[j]==uniq)):
                        total_outcome_0+=1
                
                total=df_current.value_counts()[uniq]
                percent= round((total/total_data*100),2)
                
                percentage_outcome_1=round(((total_outcome_1/total)*100),2)
                percentage_outcome_0=round((100-percentage_outcome_1),2)
                
                if(design==0):
                    print(target)
                
                print("  ",uniq,"\t\t\t\t",total,"\t",percent,"% \t\t",total_outcome_1,"\t\t\t",percentage_outcome_1,"%\t\t",total_outcome_0,"\t\t",percentage_outcome_0,"%")
                print("\n")
                design=1
        else:
            mean_value=round(df_current.mean(),2)
            
            total_abm=0
            total_blm=0
            
            total_outcome_abm_1=0
            total_outcome_abm_0=0
            
            total_outcome_blm_1=0
            total_outcome_blm_0=0
            
            for j in range(0,len(df)):
                if(df_current[j]>=mean_value):
                    total_abm+=1
                    if(df[outcome][j]==1):
                        total_outcome_abm_1+=1
                    else:
                         total_outcome_abm_0+=1
                else:
                    total_blm+=1
                    if(df[outcome][j]==1):
                        total_outcome_blm_1+=1
                    else:
                         total_outcome_blm_0+=1
           
            percent_abm=round(((total_abm/total_data)*100),2)
            percent_blm=round((100-percent_abm),2)
            
            percent_abm_1=round(((total_outcome_abm_1/total_abm)*100),2)
            percent_abm_0=round((100-percent_abm_1),2)
            
            percent_blm_1=round(((total_outcome_blm_1/total_blm)*100),2)
            percent_blm_0=round((100-percent_blm_1),2)
            
            if(design==0):
                    print(target)
                
            print(" LessThan(",mean_value,")\t\t",total_blm,"\t",percent_blm,"% \t",total_outcome_blm_1,"\t\t\t",percent_blm_1,"%\t\t\t",total_outcome_blm_0,"\t\t\t\t",percent_blm_0,"%")
            print(" GreaterThan(",mean_value,")\t\t",total_abm,"\t",percent_abm,"% \t ",total_outcome_abm_1,"\t\t\t",percent_abm_1,"%\t\t\t",total_outcome_abm_0,"\t\t\t\t",percent_abm_0,"%")
            print("\n")
            design=1
    else:
        if (df_current.nunique()<uniq_cases):
            
            for uniq in df_current.unique():
                total_outcome_1=0
                total_outcome_0=0
                for j in range(0,len(df)):
                    if((df[outcome][j]==1) and (df_current[j]==uniq)):
                        total_outcome_1+=1
                    elif((df[outcome][j]==0) and (df_current[j]==uniq)):
                        total_outcome_0+=1
                
                total=df_current.value_counts()[uniq]
                percent= round((total/total_data*100),2)
                
                percentage_outcome_1=round(((total_outcome_1/total)*100),2)
                percentage_outcome_0=round((100-percentage_outcome_1),2)
                
                if(design==0):
                    print(target)
                
                print("  ",uniq,"\t\t\t",total,"\t",percent,"% \t\t",total_outcome_1,"\t\t",percentage_outcome_1,"%\t\t\t",total_outcome_0,"\t\t\t",percentage_outcome_0,"%")
                print("\n")
                design=1
        else:
            print("Target value is string and have greater unique Values")
                

In [392]:
# With Passenger Id
Detailed_Details(df,'PassengerId','Survived')

#Doesnot seem like it really matters. PassengerId looks like in sorted manner

Comparing PassengerId  With Survived
Total 418


Column Name			Total	Percentage	Total_Outcome(1)	Percentage_Outcome(1)	Total_Outcome(0)	Percentage_Outcome(0) 
PassengerId
 LessThan( 1100.5 )		 209 	 50.0 % 	 79 			 37.8 %			 130 				 62.2 %
 GreaterThan( 1100.5 )		 209 	 50.0 % 	  73 			 34.93 %			 136 				 65.07 %




In [393]:
# With PClass
Detailed_Details(df,'Pclass','Survived')

#High class 1 have higher chance of survival. Class may matter

Comparing Pclass  With Survived
Total 418


Column Name			Total	Percentage	Total_Outcome(1)	Percentage_Outcome(1)	Total_Outcome(0)	Percentage_Outcome(0) 
Pclass
   1 				 107 	 25.6 % 		 50 			 46.73 %		 57 		 53.27 %


   2 				 93 	 22.25 % 		 30 			 32.26 %		 63 		 67.74 %


   3 				 218 	 52.15 % 		 72 			 33.03 %		 146 		 66.97 %




In [394]:
# With Name
Detailed_Details(df,'Name','Survived')

#Cannot with our model

Comparing Name  With Survived
Total 418


Column Name			Total	Percentage	Total_Outcome(1)	Percentage_Outcome(1)	Total_Outcome(0)	Percentage_Outcome(0) 
Target value is string and have greater unique Values


In [395]:
# With Sex
Detailed_Details(df,'Sex','Survived')

#Only Female survive

Comparing Sex  With Survived
Total 418


Column Name			Total	Percentage	Total_Outcome(1)	Percentage_Outcome(1)	Total_Outcome(0)	Percentage_Outcome(0) 
Sex
   male 			 266 	 63.64 % 		 0 		 0.0 %			 266 			 100.0 %


   female 			 152 	 36.36 % 		 152 		 100.0 %			 0 			 0.0 %




In [396]:
# With Age
Detailed_Details(df,'Age','Survived')

#Almost similar rate of survival. Age may not matter

Comparing Age  With Survived
Total 418


Column Name			Total	Percentage	Total_Outcome(1)	Percentage_Outcome(1)	Total_Outcome(0)	Percentage_Outcome(0) 
Age
 LessThan( 30.27 )		 200 	 47.85 % 	 77 			 38.5 %			 123 				 61.5 %
 GreaterThan( 30.27 )		 218 	 52.15 % 	  75 			 34.4 %			 143 				 65.6 %




In [397]:
#With SibSip

Detailed_Details(df,'SibSp','Survived')

#Better check correlation for this

Comparing SibSp  With Survived
Total 418


Column Name			Total	Percentage	Total_Outcome(1)	Percentage_Outcome(1)	Total_Outcome(0)	Percentage_Outcome(0) 
SibSp
   0 				 283 	 67.7 % 		 88 			 31.1 %		 195 		 68.9 %


   1 				 110 	 26.32 % 		 54 			 49.09 %		 56 		 50.91 %


   2 				 14 	 3.35 % 		 6 			 42.86 %		 8 		 57.14 %


   3 				 4 	 0.96 % 		 1 			 25.0 %		 3 		 75.0 %


   4 				 4 	 0.96 % 		 1 			 25.0 %		 3 		 75.0 %


   5 				 1 	 0.24 % 		 1 			 100.0 %		 0 		 0.0 %


   8 				 2 	 0.48 % 		 1 			 50.0 %		 1 		 50.0 %




In [398]:
#With Parch
Detailed_Details(df,'Parch','Survived')
#better to check correlation

Comparing Parch  With Survived
Total 418


Column Name			Total	Percentage	Total_Outcome(1)	Percentage_Outcome(1)	Total_Outcome(0)	Percentage_Outcome(0) 
Parch
   0 				 324 	 77.51 % 		 99 			 30.56 %		 225 		 69.44 %


   1 				 52 	 12.44 % 		 28 			 53.85 %		 24 		 46.15 %


   2 				 33 	 7.89 % 		 20 			 60.61 %		 13 		 39.39 %


   3 				 3 	 0.72 % 		 2 			 66.67 %		 1 		 33.33 %


   4 				 2 	 0.48 % 		 2 			 100.0 %		 0 		 0.0 %


   5 				 1 	 0.24 % 		 0 			 0.0 %		 1 		 100.0 %


   6 				 1 	 0.24 % 		 0 			 0.0 %		 1 		 100.0 %


   9 				 2 	 0.48 % 		 1 			 50.0 %		 1 		 50.0 %




In [399]:
#With Ticket
Detailed_Details(df,'Ticket','Survived')

Comparing Ticket  With Survived
Total 418


Column Name			Total	Percentage	Total_Outcome(1)	Percentage_Outcome(1)	Total_Outcome(0)	Percentage_Outcome(0) 
Target value is string and have greater unique Values


In [400]:
#With Fare
Detailed_Details(df,'Fare','Survived')
#High Fares have better survival rate

Comparing Fare  With Survived
Total 418


Column Name			Total	Percentage	Total_Outcome(1)	Percentage_Outcome(1)	Total_Outcome(0)	Percentage_Outcome(0) 
Fare
 LessThan( 35.63 )		 321 	 76.79 % 	 99 			 30.84 %			 222 				 69.16 %
 GreaterThan( 35.63 )		 97 	 23.21 % 	  53 			 54.64 %			 44 				 45.36 %




In [401]:
#With Embarked
Detailed_Details(df,'Embarked','Survived')
#People from  Q have more survival rate

Comparing Embarked  With Survived
Total 418


Column Name			Total	Percentage	Total_Outcome(1)	Percentage_Outcome(1)	Total_Outcome(0)	Percentage_Outcome(0) 
Embarked
   Q 			 46 	 11.0 % 		 24 		 52.17 %			 22 			 47.83 %


   S 			 270 	 64.59 % 		 88 		 32.59 %			 182 			 67.41 %


   C 			 102 	 24.4 % 		 40 		 39.22 %			 62 			 60.78 %




<h1> Now Lets Look at Feature Selection
 </h1>

In [402]:
#Looking at correlation between columns with survival

df_comp=df.iloc[:,df.columns!='Survived']
df_comp.corrwith(df['Survived'])


PassengerId   -0.023245
Pclass        -0.108615
Age           -0.000011
SibSp          0.099943
Parch          0.159120
Fare           0.191382
dtype: float64

In [403]:

df_comp=df.iloc[:,df.columns!='Survived']
df_comp.corrwith(df['Survived'])


PassengerId   -0.023245
Pclass        -0.108615
Age           -0.000011
SibSp          0.099943
Parch          0.159120
Fare           0.191382
dtype: float64

In [404]:
#dropping PassengerID , Age , SibSp 
rows=['PassengerId','Age','SibSp']
for r in rows:
    df.drop(r,axis=1,inplace=True)

In [405]:
df.columns

Index(['Pclass', 'Name', 'Sex', 'Parch', 'Ticket', 'Fare', 'Embarked',
       'Survived'],
      dtype='object')

# Checking realations for<br>
## <b>Name  , Ticket And  Sex</b>


#We cannont find Correlation between Name and Survived

#Name Cannot be Categorized 

#Possibly Survival rate cannot be determined by only names.

#So better drop it


In [406]:
df.drop('Name',axis=1,inplace=True)

In [407]:
#Tickets

df['Ticket'].sample(10)

321         2654
32     C.A. 2315
282       364859
0         330911
197       347066
343     PC 17755
96         19877
413    A.5. 3236
26        113509
357         1222
Name: Ticket, dtype: object

# Tickets some have String values and they all seem to be random.
#So lets drop it as ticket.


In [408]:
df.drop('Ticket',axis=1,inplace=True)


In [409]:
# For Sex
df['Sex'].unique()

array(['male', 'female'], dtype=object)

# Lets map Male and Female to 0s and 1s so as to compare correlation


In [410]:
df['Sex']=df['Sex'].apply(lambda x:1 if x=='female' else 0)

# Lets Check correlation


In [411]:
df['Sex'].value_counts()
df['Sex'].corr(df['Survived'])

1.0

Thats very high Correlation. So lets keep Sex as feature


In [412]:

df.columns.values

array(['Pclass', 'Sex', 'Parch', 'Fare', 'Embarked', 'Survived'],
      dtype=object)

# column names


In [413]:
col_names=df.columns.values 

for i in range(0,len(col_names)):
    uniq_datas=df[col_names[i]].unique() #unique datas of a column 
    
    if(len( uniq_datas ) < 10 ):
        print(col_names[i] ," : ", np.sort(uniq_datas))
    
    else:
        arr= df[col_names[i]].sort_values()
        chunks = len(df[col_names[i]])/100
        print( col_names[i]," : " ,np.array_split(arr,chunks)) #splitting in chunks
    

Pclass  :  [1 2 3]
Sex  :  [0 1]
Parch  :  [0 1 2 3 4 5 6 9]
Fare  :  [372    0.0000
266    0.0000
21     3.1708
133    6.4375
116    6.4375
        ...  
86     7.8792
381    7.8792
243    7.8875
60     7.8958
134    7.8958
Name: Fare, Length: 105, dtype: float64, 295     7.8958
334     7.8958
56      7.8958
71      7.8958
209     7.8958
        ...   
309    14.1083
201    14.4000
85     14.4542
249    14.4542
115    14.4542
Name: Fare, Length: 105, dtype: float64, 366    14.4583
382    14.5000
25     14.5000
344    14.5000
192    14.5000
        ...   
379    31.3875
174    31.3875
154    31.3875
338    31.5000
31     31.5000
Name: Fare, Length: 104, dtype: float64, 319     31.5000
112     31.6792
22      31.6833
194     32.5000
247     32.5000
         ...   
375    262.3750
24     262.3750
53     263.0000
69     263.0000
343    512.3292
Name: Fare, Length: 104, dtype: float64]
Embarked  :  ['C' 'Q' 'S']
Survived  :  [0 1]


# Now converting Categorical data Embarked into codes


In [414]:

embarked_df=pd.get_dummies(df['Embarked'])

df=pd.concat([df,embarked_df],axis=1)

df.drop('Embarked',axis=1,inplace=True)


# Now converting Categorical data Parch into codes


In [415]:
parched_df=pd.get_dummies(df['Parch'])

df=pd.concat([df,parched_df],axis=1)

df.drop('Parch',axis=1,inplace=True)


In [416]:
df.head()

Pclass  Sex     Fare  Survived  C  Q  S  0  1  2  3  4  5  6  9
0       3    0   7.8292         0  0  1  0  1  0  0  0  0  0  0  0
1       3    1   7.0000         1  0  0  1  1  0  0  0  0  0  0  0
2       2    0   9.6875         0  0  1  0  1  0  0  0  0  0  0  0
3       3    0   8.6625         0  0  0  1  1  0  0  0  0  0  0  0
4       3    1  12.2875         1  0  0  1  0  1  0  0  0  0  0  0

# Features Cannot be Int so Changing The Feature Names

In [417]:
for i in range(0,10):
    df.rename(columns={i:f"Parch_{i}"},inplace=True)

In [418]:
df.head()

Pclass  Sex     Fare  Survived  C  Q  S  Parch_0  Parch_1  Parch_2  \
0       3    0   7.8292         0  0  1  0        1        0        0   
1       3    1   7.0000         1  0  0  1        1        0        0   
2       2    0   9.6875         0  0  1  0        1        0        0   
3       3    0   8.6625         0  0  0  1        1        0        0   
4       3    1  12.2875         1  0  0  1        0        1        0   

   Parch_3  Parch_4  Parch_5  Parch_6  Parch_9  
0        0        0        0        0        0  
1        0        0        0        0        0  
2        0        0        0        0        0  
3        0        0        0        0        0  
4        0        0        0        0        0

In [419]:
#Checking correlations
df.corr()

Pclass       Sex      Fare  Survived         C         Q  \
Pclass    1.000000 -0.108615 -0.576619 -0.108615 -0.318543  0.252608   
Sex      -0.108615  1.000000  0.191382  1.000000  0.033684  0.115574   
Fare     -0.576619  0.191382  1.000000  0.191382  0.312040 -0.155539   
Survived -0.108615  1.000000  0.191382  1.000000  0.033684  0.115574   
C        -0.318543  0.033684  0.312040  0.033684  1.000000 -0.199786   
Q         0.252608  0.115574 -0.155539  0.115574 -0.199786  1.000000   
S         0.120783 -0.105883 -0.178464 -0.105883 -0.767375 -0.474962   
Parch_0   0.054258 -0.224158 -0.256771 -0.224158 -0.080877  0.171098   
Parch_1  -0.093179  0.136987  0.148275  0.136987  0.140263 -0.109384   
Parch_2   0.034153  0.147542  0.096940  0.147542 -0.042399 -0.102952   
Parch_3  -0.094264  0.053559  0.232155  0.053559  0.083666 -0.029898   
Parch_4  -0.021898  0.091725  0.135019  0.091725 -0.039394 -0.024382   
Parch_5   0.042775 -0.037018 -0.003723 -0.037018 -0.027822 -0.017220   
Parch_6   0.042775 -0.037018  0.009898 -0.037018 -0.027822 -0.017220   
Parch_9   0.060565  0.019655  0.042173  0.019655 -0.039394 -0.024382   

                 S   Parch_0   Parch_1   Parch_2   Parch_3   Parch_4  \
Pclass    0.120783  0.054258 -0.093179  0.034153 -0.094264 -0.021898   
Sex      -0.105883 -0.224158  0.136987  0.147542  0.053559  0.091725   
Fare     -0.178464 -0.256771  0.148275  0.096940  0.232155  0.135019   
Survived -0.105883 -0.224158  0.136987  0.147542  0.053559  0.091725   
C        -0.767375 -0.080877  0.140263 -0.042399  0.083666 -0.039394   
Q        -0.474962  0.171098 -0.109384 -0.102952 -0.029898 -0.024382   
S         1.000000 -0.039328 -0.054392  0.105450 -0.055576  0.051335   
Parch_0  -0.039328  1.000000 -0.699793 -0.543545 -0.157850 -0.128729   
Parch_1  -0.054392 -0.699793  1.000000 -0.110354 -0.032048 -0.026135   
Parch_2   0.105450 -0.543545 -0.110354  1.000000 -0.024892 -0.020300   
Parch_3  -0.055576 -0.157850 -0.032048 -0.024892  1.000000 -0.005895   
Parch_4   0.051335 -0.128729 -0.026135 -0.020300 -0.005895  1.000000   
Parch_5   0.036256 -0.090916 -0.018458 -0.014337 -0.004164 -0.003395   
Parch_6   0.036256 -0.090916 -0.018458 -0.014337 -0.004164 -0.003395   
Parch_9   0.051335 -0.128729 -0.026135 -0.020300 -0.005895 -0.004808   

           Parch_5   Parch_6   Parch_9  
Pclass    0.042775  0.042775  0.060565  
Sex      -0.037018 -0.037018  0.019655  
Fare     -0.003723  0.009898  0.042173  
Survived -0.037018 -0.037018  0.019655  
C        -0.027822 -0.027822 -0.039394  
Q        -0.017220 -0.017220 -0.024382  
S         0.036256  0.036256  0.051335  
Parch_0  -0.090916 -0.090916 -0.128729  
Parch_1  -0.018458 -0.018458 -0.026135  
Parch_2  -0.014337 -0.014337 -0.020300  
Parch_3  -0.004164 -0.004164 -0.005895  
Parch_4  -0.003395 -0.003395 -0.004808  
Parch_5   1.000000 -0.002398 -0.003395  
Parch_6  -0.002398  1.000000 -0.003395  
Parch_9  -0.003395 -0.003395  1.000000

# Here Lets look between Pclass and Fare


In [420]:

df['Pclass'].corr(df['Fare'])

-0.5766190772528161

# A close correlation. Maybe we can discard a feature.
# checking between Survived


In [421]:

print("With Pclass")
print(df['Pclass'].corr(df['Survived']))

print("With Fare")
print(df['Fare'].corr(df['Survived']))

With Pclass
-0.10861452091523417
With Fare
0.19138247897510127


# Closer relation with Fare so lets drop PClass


In [422]:
df.drop('Pclass',axis=1,inplace=True)

In [423]:
df.head()

Sex     Fare  Survived  C  Q  S  Parch_0  Parch_1  Parch_2  Parch_3  \
0    0   7.8292         0  0  1  0        1        0        0        0   
1    1   7.0000         1  0  0  1        1        0        0        0   
2    0   9.6875         0  0  1  0        1        0        0        0   
3    0   8.6625         0  0  0  1        1        0        0        0   
4    1  12.2875         1  0  0  1        0        1        0        0   

   Parch_4  Parch_5  Parch_6  Parch_9  
0        0        0        0        0  
1        0        0        0        0  
2        0        0        0        0  
3        0        0        0        0  
4        0        0        0        0

<h1> Now Build Model </h1>

In [424]:
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score
from sklearn.model_selection import GridSearchCV

In [425]:
#Now Finally Implementing Machine Learning
svc=SVC()
knn=KNeighborsClassifier()
dtc=DecisionTreeClassifier()
nb=GaussianNB()

In [426]:
models={'svc':svc,'knn':knn,'dtc':dtc,'nb':nb}

In [427]:
x,y=df.drop('Survived',axis=1),df['Survived']

In [428]:
train_x,test_x,train_y,test_y=train_test_split(x,y,test_size=0.2,random_state=42)

In [429]:
model_accuracy={}

for model in models.keys():
    train_x,test_x,train_y,test_y=train_test_split(x,y,test_size=0.2,random_state=500)
    mod = models[model].fit(train_x,train_y)
    predict=mod.predict(test_x)
    #print(f"{model}",classification_report(predict,test_y))
    print(f"{model}")
    print(confusion_matrix(predict,test_y))
    
    if model not in model_accuracy.keys():
        model_accuracy.update({model:accuracy_score(test_y,predict)})

svc
[[55 27]
 [ 1  1]]
knn
[[54  5]
 [ 2 23]]
dtc
[[56  0]
 [ 0 28]]
nb
[[56  0]
 [ 0 28]]


In [430]:
model_accuracy

{'svc': 0.6666666666666666, 'knn': 0.9166666666666666, 'dtc': 1.0, 'nb': 1.0}

# Lets try With GridSearch


In [431]:
params={
    'knn': {'n_neighbors':[3,5,7,9,11,13,15],
    'metric':['cosine','euclidean','manhattan'],
    'weights':['uniform','distance']},
    
    'svc': {
    'C':[0.1,1,10,100], 
    'gamma':[1,0.1,0.01,0.01], 
    'kernel':['rbf','linear']},
    
    'dtc':{
        'criterion':['gini','entropy'],
        'max_depth':[2,4,6,8,10,12]
    },
    
    'nb':{
        'priors': [None],
    'var_smoothing': [0.00000001, 0.000000001, 0.00000001]
    }
    
    }

In [432]:
model_accuracy={}
for model in models.keys():
    mod = GridSearchCV(
    models[model],
    params[model],
    verbose=1, #Progress bar showing
    cv=20, #cross validation
    n_jobs=-1, #cores to assign
)
    gridsearch_result=mod.fit(train_x,train_y)
    predict=mod.predict(test_x)
    print(f"{model} : ",gridsearch_result.best_estimator_)
    print(f"{model}")
    print(confusion_matrix(predict,test_y))
    if model not in model_accuracy.keys():
        model_accuracy.update({model:accuracy_score(test_y,predict)})

Fitting 20 folds for each of 32 candidates, totalling 640 fits
svc :  SVC(C=0.1, gamma=1, kernel='linear')
svc
[[56  0]
 [ 0 28]]
Fitting 20 folds for each of 42 candidates, totalling 840 fits
knn :  KNeighborsClassifier(metric='cosine', n_neighbors=3, weights='distance')
knn
[[56  1]
 [ 0 27]]
Fitting 20 folds for each of 12 candidates, totalling 240 fits
dtc :  DecisionTreeClassifier(max_depth=2)
dtc
[[56  0]
 [ 0 28]]
Fitting 20 folds for each of 3 candidates, totalling 60 fits
nb :  GaussianNB(var_smoothing=1e-08)
nb
[[56  0]
 [ 0 28]]


In [372]:
model_accuracy

{'svc': 1.0, 'knn': 0.9880952380952381, 'dtc': 1.0, 'nb': 1.0}

In [50]:
#Maybe the result is nice.